### PR TITLE
Add dirs

### PR DESCRIPTION
**Pull request by ajeddeloh**
WIP. The filesystemEntries in internal/exec/stages/files/files.go seems gross, but go's lack of generics makes it hard to find a better solution without duplicating a ton of code.

### CODE REVIEW

**Comment by ajeddeloh** [comment]
Cleanup up commits

**Comment by crawford** [comment]
Please clean up comments per Go guidelines.

File: config/types/directory.go [hunk 1]
```diff
@@ -0,0 +1,45 @@
...
+	count := 0
+	for p != "/" {
+		count++
+		p = filepath.Dir(p)
+	}
+	return count
+}
+
+// Needed to sort directories so /foo get created before /foo/bar if they are both specified
+type ByPathSections []Directory
```
Comment (crawford) [inline config/types/directory.go]: DirectorySegments

Comment (crawford) [inline config/types/directory.go]: Oh, this is not what I thought this was for...

I would just move this into the single file that uses it.

File: doc/configuration.md [hunk 2]
```diff
@@ -50,6 +50,14 @@ The Ignition configuration is a JSON document conforming to the following specif
       * **_id_** (integer): the user ID of the owner.
     * **_group_** (object): specifies the group of the owner.
       * **_id_** (integer): the group ID of the owner.
+  * **_directories_** (list of objects): the list of directories, rooted in this particular filesystem, to be created.
```
Comment (crawford) [inline doc/configuration.md]: "rooted in this particular filesystem"

Which filesystem?

File: doc/configuration.md [hunk 3]
```diff
@@ -50,6 +50,14 @@ The Ignition configuration is a JSON document conforming to the following specif
...
+    * **filesystem** (string): the internal identifier of the filesystem. This matches the last filesystem with the given identifier.
+    * **path** (string): the absolute path to the directory.
+    * **_mode_** (integer): the directory's permission mode. Note that the mode must be properly specified as a **decimal** value (i.e. 0644 -> 420).
```
Comment (crawford) [inline doc/configuration.md]: Can you use a better example? This is an odd mode for a directory.

File: config/types/directory.go (hunk 1 above)
Comment (ajeddeloh) [inline config/types/directory.go]: The NumSegments needs to stay in this package if it is to stay a method since you can't define methods on types outside the current package, unless you make it a function that takes the Directory as a parameter instead or make a type alias. The ByDirectorySegments can be moved out though. Which were you referring to?

Comment (crawford) [inline config/types/directory.go]: ByPathSections

File: config/types/directory.go [hunk 4]
```diff
@@ -0,0 +1,45 @@
...
+package types
+
+import (
+	"path/filepath"
+)
+
+// Just a special case of File
+type Directory File
+
+func (d *Directory) NumSegments() int {
```
Comment (crawford) [inline config/types/directory.go]: I would call this Depth.

File: doc/configuration.md (hunk 2 above)
Comment (ajeddeloh) [inline doc/configuration.md]: The one it lists. 
I used the files section as a base for this. Admittedly both seem a little unclear; it looks like it was meant for the v1 spec. Change both to omit "rooted in this particular filesystem" and make the filesystem description include something along those lines?

**Comment by ajeddeloh** [comment]
Fixed. Also added owner/group which I forgot before.

File: doc/configuration.md [hunk 5]
```diff
@@ -50,6 +50,14 @@ The Ignition configuration is a JSON document conforming to the following specif
       * **_id_** (integer): the user ID of the owner.
     * **_group_** (object): specifies the group of the owner.
       * **_id_** (integer): the group ID of the owner.
+  * **_directories_** (list of objects): the list of directories to be created.
+    * **filesystem** (string): the internal identifier of the filesystem to create the directory in. This matches the last filesystem with the given identifier.
```
Comment (crawford) [inline doc/configuration.md]: Passive voice. "the internal identifier of the filesystem in which to create the directory."

File: config/translate.go [hunk 6]
```diff
@@ -82,18 +82,20 @@ func TranslateFromV1(old v1.Config) (types.Config, error) {
 		config.Storage.Filesystems = append(config.Storage.Filesystems, filesystem)
 
 		for _, oldFile := range oldFilesystem.Files {
-			file := types.File{
-				Filesystem: filesystem.Name,
-				Path:       types.Path(oldFile.Path),
+			file := types.RegFile{
```
Comment (vcaputo) [inline config/translate.go]: I feel like the abbreviation of `RegularFile` is somewhat spurious, when the `Directory` type hasn't been abbreviated, and `RegularFile` is only two characters longer.  A casual inexperienced reader may interpret `RegFile` as register file or something unexpected, just use the expanded form, it's more clear.

File: config/types/file.go [hunk 7]
```diff
@@ -24,13 +24,18 @@ var (
 	ErrFileIllegalMode = errors.New("illegal file mode")
 )
 
+// File represents all common info for files (special types, e.g. directories, included).
```
Comment (vcaputo) [inline config/types/file.go]: `Node` is more appropriate than `File` for this, since it very much relates to the `inode`-associated context.

Probably makes sense to `s/File(Mode|User|Group)/Node\1/g` as well...

**Comment by vcaputo** [comment]
minor nomenclature nitpicks, otherwise lgtm.

**Comment by ajeddeloh** [comment]
Fixed nomenclature

File: config/types/file.go [hunk 8]
```diff
@@ -24,20 +24,26 @@ var (
 	ErrFileIllegalMode = errors.New("illegal file mode")
 )
 
+// Node represents all common info for files (special types, e.g. directories, included).
+type Node struct {
```
Comment (vcaputo) [inline config/types/file.go]: Does it seem like all the Node* stuff should move into their own `node.go`?

**Comment by vcaputo** [comment]
other than maybe putting the Node* stuff into their own listing, which is up to you, lgtm.

**Comment by crawford** [comment]
Commit message for 08ed373c should read "config: add directories".

File: config/types/directory.go [hunk 9]
```diff
@@ -0,0 +1,32 @@
...
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package types
+
+import (
+	"path/filepath"
+)
+
+// Directory is just a special case of Node.
```
Comment (crawford) [inline config/types/directory.go]: This comment isn't very helpful. Can you elaborate a bit or just remove it?

File: config/types/directory.go [hunk 10]
```diff
@@ -0,0 +1,32 @@
...
+type Directory Node
+
+func (d *Directory) Depth() int {
+	p := filepath.Clean(string(d.Path))
```
Comment (crawford) [inline config/types/directory.go]:
```go
	count := 0
	for p := filepath.Clean(d); p != "/"; count++ {
		p = filepath.Dir(p)
	}
	return count
```

File: internal/exec/stages/files/files.go [hunk 11]
```diff
@@ -79,57 +80,132 @@ func (s stage) Run(config types.Config) bool {
...
+
+func (lst ByDirectorySegments) Swap(i, j int) {
+	lst[i], lst[j] = lst[j], lst[i]
+}
+
+func (lst ByDirectorySegments) Less(i, j int) bool {
+	return lst[i].Depth() < lst[j].Depth()
+}
+
+// mapEntriesToFilesystems builds a map of filesystems to files. If multiple
```
Comment (crawford) [inline internal/exec/stages/files/files.go]: Can you also mention that directories are sorted?

File: internal/exec/stages/files/files.go [hunk 12]
```diff
@@ -79,57 +80,132 @@ func (s stage) Run(config types.Config) bool {
...
+			entryMap[fs] = append(entryMap[fs], dirEntry(d))
 		} else {
-			s.Logger.Crit("the filesystem (%q), was not defined", fsn)
+			s.Logger.Crit("the filesystem (%q), was not defined", d.Filesystem)
 			return nil, ErrFilesystemUndefined
 		}
 	}
 
-	return fileMap, nil
+	for _, f := range config.Storage.Files {
```
Comment (crawford) [inline internal/exec/stages/files/files.go]: Duplicate. Can you make this more general?

**Comment by crawford** [comment]
Can you combine 5e116df and 4b3e5ca?

File: internal/exec/stages/files/files.go (hunk 12 above)
Comment (ajeddeloh) [inline internal/exec/stages/files/files.go]: I don't think so without sacraficing type safety. Because you can't convert concrete types in go, even to/from an embedded struct, you can't write a generic "add `Node`-like object" function to add them to entryMap without it taking an interface{}. I figured the code duplication was worth the type safety. Additionally, you need the dual for loops anyway since you need to loop over config.Storage.{Files,Directories} and create either fileEntry or dirEntry depending on which one it is.

File: config/types/directory.go (hunk 9 above)
Comment (ajeddeloh) [inline config/types/directory.go]: removed

File: config/types/directory.go (hunk 10 above)
Comment (ajeddeloh) [inline config/types/directory.go]: done.

File: internal/exec/stages/files/files.go (hunk 11 above)
Comment (ajeddeloh) [inline internal/exec/stages/files/files.go]: done.

**Comment by ajeddeloh** [comment]
combined.

File: internal/exec/stages/files/files.go [hunk 13]
```diff
@@ -79,57 +80,132 @@ func (s stage) Run(config types.Config) bool {
...
+	return nil
+}
+
+type dirEntry types.Directory
+
+func (tmp dirEntry) create(l *log.Logger, u util.Util) error {
+	d := types.Directory(tmp)
+	err := l.LogOp(func() error {
+		path := u.JoinPath(string(d.Path))
+		if err := util.MkdirForFile(path); err != nil {
```
Comment (crawford) [inline internal/exec/stages/files/files.go]: Why is this needed?

File: internal/exec/stages/files/files.go [hunk 14]
```diff
@@ -79,57 +80,132 @@ func (s stage) Run(config types.Config) bool {
...
+			return err
+		}
+
+		if err := os.Mkdir(path, os.FileMode(d.Mode)); err != nil {
```
Comment (crawford) [inline internal/exec/stages/files/files.go]: MkdirAll?

File: internal/exec/stages/files/files.go [hunk 15]
```diff
@@ -79,57 +80,132 @@ func (s stage) Run(config types.Config) bool {
...
+			return err
+		}
+
+		// Set the ownership and mode are as requested (since Mkdir doesn't set ownership and can be affect by the sticky bit)
```
Comment (crawford) [inline internal/exec/stages/files/files.go]: "Mkdir doesn't set ownership" I don't quite follow this. Is this a bug in Go? /cc @vcaputo

File: internal/exec/stages/files/files.go (hunk 14 above)
Comment (ajeddeloh) [inline internal/exec/stages/files/files.go]: MkdirAll uses the same permissions for all directories. IIRC we wanted default permissions for all directories leading up to the one being created.

File: internal/exec/stages/files/files.go (hunk 15 above)
Comment (ajeddeloh) [inline internal/exec/stages/files/files.go]: `d.Filemode` only has permission information, thus we need to apply owner/group seperately.

Comment (crawford) [inline internal/exec/stages/files/files.go]: Sorry, I copied the wrong snippet. Why do we need to call Chmod?

Comment (ajeddeloh) [inline internal/exec/stages/files/files.go]: So I went looking in the go source when I saw this: https://github.com/coreos/ignition/blob/master/internal/exec/util/file.go#L195 and the implementation for os.Mkdir and os.OpenFile. They both seem to handle the sticky bit case, but it was obviously an issue in the file version, so I figured I better cover the case in the directory version as well.

File: internal/exec/stages/files/files.go (hunk 14 above)
Comment (crawford) [inline internal/exec/stages/files/files.go]: Let's create all of the subdirectories with the same permissions. I think that's going to be simpler for the user and a more likely scenario.

File: internal/exec/stages/files/files.go (hunk 15 above)
Comment (vcaputo) [inline internal/exec/stages/files/files.go]: @crawford is it still unclear?  it doesn't set ownership... like the comment says :)

Comment (crawford) [inline internal/exec/stages/files/files.go]: I'm confused about the mode. How is that call any different from what Mkdir is doing internally?

File: internal/exec/stages/files/files.go (hunk 14 above)
Comment (ajeddeloh) [inline internal/exec/stages/files/files.go]: What about ownership? I think if we do the permissions, I think we should also do owner/group.

Comment (crawford) [inline internal/exec/stages/files/files.go]: Agreed.

File: internal/exec/stages/files/files.go (hunk 15 above)
Comment (ajeddeloh) [inline internal/exec/stages/files/files.go]: That's what I was referring to with the os.Mkdir and os.OpenFile implementions. looking at the util/file.go comment, it appears there were issues with the sticky bit affecting it. I figured since they have the same sticky bit checks in their implementations, if we had an issue with OpenFile, we would probably have the same issue with Mkdir and thus we should have the extra chmod to ensure the mode was set correctly.

Comment (ajeddeloh) [inline internal/exec/stages/files/files.go]: Interestingly, all I could ever find was issues relating to solaris/bsd, not linux.

Comment (vcaputo) [inline internal/exec/stages/files/files.go]: The mkdir syscall will apply umask to the mode, chmod is explicit.  I was likely being conservative when I wrote that code originally, we could just take care to clear the umask of the executing environment for ignition, but manual testing would then potentially produce different results if the umask was something else.

Setting the mode explicitly takes umask out of the equation...

File: internal/exec/stages/files/files.go (hunk 14 above)
Comment (ajeddeloh) [inline internal/exec/stages/files/files.go]: Fixed.

**Comment by crawford** [comment]
This looks good to me. Can you rebase onto spec one last time?
